contributors/design-proposals/storage/container-storage-interface-node-status.md

```diff
@@ -0,0 +1,105 @@
+# Node.Status usage for CSI
+
+***Version:*** Alpha
+
+***Author:*** @jsafrane
+
+## Terminology
+
+Term | Definition
+---|---
+Container Storage Interface (CSI) | A specification attempting to establish an industry standard interface that Container Orchestration Systems (COs) can use to expose arbitrary storage systems to their containerized workloads.
+CSI Volume Plugin | A in-tree volume plugin that acts as an adapter and enables out-of-tree, third-party CSI volume drivers to be used in Kubernetes.
+CSI Volume Driver | An out-of-tree CSI compatible implementation of a volume plugin that can be used in Kubernetes through the Kubernetes CSI Volume Plugin.
+
+## Background & Motivations
+
+Storage backends may use different node IDs than Kubernetes - they may run on a distinct network with their own IP addresses or use separate DNS domains. Therefore when Kubernetes talks to a CSI Volume Driver, it must use node ID that is understandable to the CSI Volume Driver. A CSI Volume Driver reports ID of a node where it runs during registration of the driver to kubelet. Kubelet then collects node IDs of all CSI drivers installed on the node and currently publishes it as JSON dictionary in `csi.volume.kubernetes.io/nodeid`. In the end, each node has node IDs for all CSI drivers that run on the node.
+
+Example of a 1.11 node running hypothetical CSI drivers for AWS and Gluster:
+
+```yaml
+apiVersion: v1
+kind: Node
+metadata:
+  name: node1.example.com
+  annotations:
+    csi.volume.kubernetes.io/nodeid: { \"csi.amazon.com/ebs\": \"ip-172-31-60-59.ec2.internal\", \"csi.gluster.org\": \"172.31.60.59\" }
+```
+
+This annotation cannot be extended with additional information as CSI specification evolves. In this proposal, we want to:
+
+  * Move this annotation into proper Node field.
+  * Add topology information introduced in CSI v0.3.0.
+
+This is part of
+
+* https://github.com/kubernetes/features/issues/557
+* https://github.com/kubernetes/community/pull/2034
+
+This fixes:
+
+* https://github.com/kubernetes/kubernetes/issues/66497
+
+
+## API
+
+```go
+// NodeStatus is information about the current status of a node.
+type NodeStatus struct {
+    // ...
+
+    // List of CSI drivers running on the node and their properties.
+    CSIDrivers []CSIDriverStatus
+}
+
+type CSIDriverStatus struct {
```

Comment:
Since this was named Status I was expecting some field representing the health of the CSI driver. Should there be a status or health property?